### ISSUE TITLE
Use new Load() overloads when only a vector is required

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#921]
There are many places in command-line programs (e.g. `logistic_regression_main.cpp` and others) where we have code like

```
arma::Mat<size_t> labelsIn;
data::Load(labelsFile, labelsIn, true);
// convert this to an arma::Row<size_t> or whatever
```

It would be much better if we dispensed with this step and simply loaded into the correct Armadillo type directly, and this would allow us to reduce the size of the codebase a little bit.  A PR to fix this issue should handle all occurrences of this pattern in the mlpack codebase.  But a PR can't be merged until #920 is fixed---we need to ensure that a user, when they pass an invalid labels file or something, gets an exception so they know something is wrong.

**Comment by ShihaoJing** [comment]
@rcurtin I'd like to take this.

**Comment by ShihaoJing** [comment]
I noticed that in `logistic_regression_main.cpp` and others, lables can loaded this way:
```
responsesMat = std::move(CLI::GetParam<arma::Mat<size_t>>("labels"));
if (responsesMat.n_cols == 1)
      responses = responsesMat.col(0).t();
    else
      responses = responsesMat.row(0);
```
When I tried to change parma loading to 
`responsesMat = std::move(CLI::GetParam<arma::Col<size_t>>("labels"));`
It gives me some errors:
```
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/util/param_data_impl.hpp:70:5: error: no matching function for call to 'Load'
    data::Load(value, matrix, true, !d.noTranspose);
    ^~~~~~~~~~
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/util/cli_impl.hpp:194:16: note: in instantiation of function template specialization
      'mlpack::util::HandleParameter<arma::Col<double> >' requested here
  return util::HandleParameter<T>(v, d);
               ^
/Users/shihaojing/Projects/mlpack/src/mlpack/methods/linear_regression/linear_regression_main.cpp:143:34: note: in instantiation of function template
      specialization 'mlpack::CLI::GetParam<arma::Col<double> >' requested here
      responses = std::move(CLI::GetParam<vec>("training_responses"));
                                 ^
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/data/load_impl.hpp:112:6: note: candidate function [with eT = double] not viable: no known
      conversion from 'typename ParameterType<Col<double> >::type' (aka 'arma::Col<double>') to 'const std::string' (aka 'const basic_string<char,
      char_traits<char>, allocator<char> >') for 1st argument
bool Load(const std::string& filename,
     ^
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/data/load_impl.hpp:525:6: note: candidate function [with T = bool] not viable: no known
      conversion from 'typename ParameterType<Col<double> >::type' (aka 'arma::Col<double>') to 'const std::string' (aka 'const basic_string<char,
      char_traits<char>, allocator<char> >') for 1st argument
bool Load(const std::string& filename,
     ^
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/data/load_impl.hpp:361:6: note: candidate template ignored: could not match
      'DatasetMapper<type-parameter-0-1>' against 'bool'
bool Load(const std::string& filename,
     ^
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/data/load_impl.hpp:95:6: note: candidate function template not viable: requires at most 3
      arguments, but 4 were provided
bool Load(const std::string& filename,
     ^
/Users/shihaojing/Projects/mlpack/src/mlpack/../mlpack/core/data/load_impl.hpp:104:6: note: candidate function template not viable: requires at most 3
      arguments, but 4 were provided
bool Load(const std::string& filename,
```
I have pulled #918 and there should be a load function that can load vector directly.

**Comment by rcurtin** [comment]
Ah, yes, so this one actually does turn out to be somewhat more complicated than I had thought originally.  It turns out, we need to get our hands dirty and make the following modifications:

- Add a new `PARAM_VECTOR()` and `PARAM_UVECTOR()` macro, like the others in `src/mlpack/core/util/param.hpp`.
- Modify `output_param_impl.hpp` and `param_data_impl.hpp` to correctly handle loading/saving vectors instead of matrices:
  - Look at the overload of `OutputParam()` that handles Armadillo objects; we'll have to add some logic to handle when the Armadillo object is a vector and then call the correct overload of `data::Save()`.
  - Look at the overload of `HandleParameter()` that handles Armadillo objects; we'll have to add some logic to handle when the Armadillo object is a vector and then call the correct overload of `data::Load()`.
- Adapt `*_main.cpp` program to use `PARAM_UVECTOR()` for labels instead of `PARAM_UMATRIX()`.

I know, maybe that is more involved than what you signed up for, but it might be fun to investigate how the CLI interface works.  If the code there doesn't make sense, I would suggest reading about SFINAE and templates, and it should be more clear.

**Comment by ShihaoJing** [comment]
@rcurtin This is a good start point for me to investigate how the CLI interface works, thank!

**Comment by rcurtin** [comment]
With #927 merged and 8bd7f79 this is now fixed.  Thanks!
